hail/python/hailtop/aiotools/tasks.py

```diff
@@ -0,0 +1,22 @@
+import asyncio
+import logging
+import weakref
+
+
+log = logging.getLogger('aiotools.tasks')
+
+
+class BackgroundTaskManager:
+    def __init__(self):
+        self.tasks: weakref.WeakSet = weakref.WeakSet()
+
+    def ensure_future(self, coroutine):
+        self.tasks.add(asyncio.ensure_future(coroutine))
+
+    def shutdown(self):
+        for task in self.tasks:
+            try:
+                if task:
```

Comment:
Should this check whether the task is done rather than exists?

Reply:
Oh whoops. This is always True because tasks are always true. No, it shouldn't check if the task is done, shutdown is only called when the app needs to go away *now*. Anything that needs to avoid getting cancelled can use `asyncio.shield`. Those tasks will continue to keep the app running.